### PR TITLE
replace HistoricalValidationCode usages with ValidationCodeByHash

### LINKED ISSUE #3188: Deprecate historical-validation-code runtime API and switch to using code-by-hash
As seen in https://github.com/paritytech/polkadot/pull/3187 , the `historical_validation_code` runtime API is not robust and has many edge-cases. 

For all the use-cases we'd need with disputes and approval checking, `code_by_hash` is sufficient. The `historical_validation_code` runtime API should be deprecated and uses changed in favor of `code_by_hash`.

### PR DESCRIPTION
**Pull request by andresilva**
Remove all code related to historical validation code and replace usages of it with the validation code by hash API. Fix #3188.

### CODE REVIEW

File: node/core/runtime-api/src/tests.rs [hunk 1]
```diff
@@ -41,7 +41,7 @@ struct MockRuntimeApi {
     session_index_for_child: SessionIndex,
     session_info: HashMap<SessionIndex, SessionInfo>,
     validation_code: HashMap<ParaId, ValidationCode>,
-    historical_validation_code: HashMap<ParaId, Vec<(BlockNumber, ValidationCode)>>,
+	validation_code_by_hash: HashMap<ValidationCodeHash, ValidationCode>,
```
Comment (ordian) [inline node/core/runtime-api/src/tests.rs:44]: looks like #3206 introduced some spaces, could you fmt this file with tabs?

File: parachain/src/primitives.rs [hunk 2]
```diff
@@ -62,6 +62,9 @@ impl ValidationCode {
 	}
 }
 
+/// A hash of the parachain validation code.
+pub type ValidationCodeHash = Hash;
```
Comment (ordian) [inline parachain/src/primitives.rs:66]: I'd prefer the newtype pattern for more type safety, although it's more boilerplate as well.

**Review by ordian — APPROVED** [review]
Looks straightforward.

**Review by rphmeier — APPROVED** [review]
Thanks!

File: parachain/src/primitives.rs (hunk 2 above)
Comment (andresilva) [reply parachain/src/primitives.rs:66]: Agreed, I have the changes ready, but they're touching a lot of stuff, probably easier to review if I make them on a follow-up PR.

**Comment by rphmeier** [comment]
bot merge

**Pull request merged**